### PR TITLE
Nova services adoption (no extra cell, no ffu)

### PR DESCRIPTION
**Pull request by bogdando**
Extracted nova services adoption w/o FFU being done from https://github.com/openstack-k8s-operators/data-plane-adoption/pull/176

Note about remapping cell DB names from OSP cells naming scheme
to the NG scheme with the superconductor layout.

Add a step to rename default cell as cell1, and to delete stale
Nova services records from cell1 DB during initial databases import,
to properly transition it into a superconductor layout later on.

Adjust minor gaps in the Placement adoption docs

Address the switch for service overrides spec instead of
externalEndpoints, where it is missing on the path to Nova adotpion.

Remove Nova Metadata secret creation workarounds from the EDPM
adotopion docs and test suits.

Provide workaround for renaming 'default' cell's DB during adoption.

Add test suits for Nova CP services adoption.

Add missing nova and libvirt services for the edpm adoption tests.
Add nova extra config custom service to complete pre-FFU deployment.
Use nova-extra-config service instead of the nova service,
because by today design choices, each config change requires a new
service deployed with ansible EDPM framework.

Document/automate creating a Nova migration key-pair requiremed
for EDPM deploy.

Stop Nova services in stop_openstack_services instead of edpm_adoption
(that was too late to do that).

Jira: [OSPRH-338](https://issues.redhat.com/browse/OSPRH-338)

### CODE REVIEW

**Comment by bogdando** [comment]
@jistr @gibi it fails as there is no https://github.com/openstack-k8s-operators/nova-operator/pull/566 in the CI builds. I'm hitting on my dev env this for the same reason. How could we improve the delivery of fixes into builds?..

**Comment by bogdando** [comment]
this passes testing on my dev env, if I manually patch CSVs of nova and openstack operators to latest tags. We really need to fix CI...

**Comment by gibizer** [comment]
The CI job failed as the cell0 conductor was not able to talk to keystone.
```
2023-11-08 18:26:12.128 1 ERROR nova keystoneauth1.exceptions.http.Unauthorized: The request you have made requires authentication. (HTTP 401) (Request-ID: req-af061dbb-93a9-41e6-b5ca-4f3aaf248003)
```
https://review.rdoproject.org/zuul/build/c1d1eed1ce8b411e9a3975de4b6c4dd0/log/controller/pod/nova-cell0-conductor-0-logs.txt

**Comment by bogdando** [comment]
> The CI job failed as the cell0 conductor was not able to talk to keystone.
> 
> ```
> 2023-11-08 18:26:12.128 1 ERROR nova keystoneauth1.exceptions.http.Unauthorized: The request you have made requires authentication. (HTTP 401) (Request-ID: req-af061dbb-93a9-41e6-b5ca-4f3aaf248003)
> ```
> 
> https://review.rdoproject.org/zuul/build/c1d1eed1ce8b411e9a3975de4b6c4dd0/log/controller/pod/nova-cell0-conductor-0-logs.txt

this is really weird, I was hitting exactly the same issue on my env, until I retried it from scratch, after ``make crc_storage_cleanup crc_storate``.

**Comment by bogdando** [comment]
recheck

**Comment by bogdando** [comment]
recheck

**Comment by bogdando** [comment]
@gibizer @jistr this passes now. Although on my env I'm hitting ovn adoption problems with
```
"ovsdb-client: failed to connect to \"tcp:172.17.0.30:6641\" (Connection refused)"]
```
if it works for you, please consider merging

File: docs/openstack/stop_openstack_services.md [hunk 1]
```diff
@@ -64,6 +68,7 @@ there are no ongoing  operations.
 1- Connect to all the controller nodes.
 2- Stop the services.
 3- Make sure all the services are stopped.
+4- Repeat steps 1-3 for compute hosts (workloads running on dataplane will not be affected)
```
Comment (jistr) [inline docs/openstack/stop_openstack_services.md:71]: If we want to stop those services everywhere, we should probably expand the snippet below to iterate over compute nodes too and not just `CONTROLLER${i}_SSH`. Not a blocker but should be addressed in a follow-up.

File: tests/roles/stop_openstack_services/tasks/main.yaml [hunk 2]
```diff
@@ -6,6 +6,7 @@
       CONTROLLER2_SSH="{{ controller2_ssh }}"
       CONTROLLER3_SSH="{{ controller3_ssh }}"
 
+# TODO(bogdando): split Compute services on dataplane after the multi-node/multi-cells adoption supported
```
Comment (jistr) [inline tests/roles/stop_openstack_services/tasks/main.yaml:9]: Indeed, the below works thanks to being Standalone where nova-compute is collocated with the controller services, but the snippet in the docs and the one below ignores compute nodes. We probably need to introduce COMPUTE${i}_SSH variables. The OVN adoption procedure also SSH's into the compute node and should be eventually able to work with multiple computes, so this issue is not specific to Nova currently. Not a blocker.

**Review by jistr — APPROVED** [review]

**Review by jistr — APPROVED** [review]
:ship:

File: tests/roles/stop_openstack_services/tasks/main.yaml (hunk 2 above)
Comment (SeanMooney) [reply tests/roles/stop_openstack_services/tasks/main.yaml:9]: stoping the nova/libvirt service is somthing i think can and should be done in the relevant edpm roles yes

just be aware that we also run nova-compute with ironic on the controller and in that case the controller will not be an edpm compute as that nova-compute will be moving to openshift.

Comment (jistr) [reply tests/roles/stop_openstack_services/tasks/main.yaml:9]: At that time the compute services are pointed at the old RabbitMQ. We aren't stopping Rabbit right now but i wonder if we should. Also we are stopping the Nova controller services, so even if the compute services do make RPC calls and Rabbit is running, there's no service running that could alter DBs i think.

File: docs/openstack/nova_adoption.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  oc get novaconductors --field-selector metadata.name=nova-cell0-conductor -o jsonpath='{.items[0].status.conditions}' \
+  | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+  oc get novaconductors --field-selector metadata.name=nova-cell1-conductor -o jsonpath='{.items[0].status.conditions}' \
+  | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+  oc get novametadata --field-selector metadata.name=nova-metadata -o jsonpath='{.items[0].status.conditions}' \
+  | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+  oc get novanovncproxies --field-selector metadata.name=nova-cell1-novncproxy -o jsonpath='{.items[0].status.conditions}' \
+  | jq -e '.[]|select(.type=="Ready" and .status=="True")'
+  oc get novaschedulers --field-selector metadata.name=nova-scheduler -o jsonpath='{.items[0].status.conditions}' \
+  | jq -e '.[]|select(.type=="Ready" and .status=="True")'
```
Comment (gibizer) [inline docs/openstack/nova_adoption.md:128]: I would simplify this to:
```
oc wait --for condition=Ready --timeout=300s Nova/nova
```
If it is not enough to watch for the Ready condition of the Nova CR then that is a bug and please file it to nova-operator.

File: docs/openstack/edpm_adoption.md [hunk 4]
```diff
@@ -160,6 +162,9 @@ done
       - install-os
       - configure-os
       - run-os
+      - libvirt
+      - nova
+      - nova-compute-extraconfig
```
Comment (gibizer) [inline docs/openstack/edpm_adoption.md:167]: The `nova-compute-extraconfig` is enough we don't need two nova ansible runs in the same deployment. The run of the second overwrites  anything that was done by the first.

File: tests/roles/dataplane_adoption/tasks/main.yaml [hunk 5]
```diff
@@ -183,6 +174,8 @@
         - install-os
         - configure-os
         - run-os
+        - libvirt
+        - nova
```
Comment (gibizer) [inline tests/roles/dataplane_adoption/tasks/main.yaml:178]: This does not seems to use the nova-compute-extraconfig created above.

File: tests/roles/nova_adoption/tasks/main.yaml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    file: wait.yaml
+
+# TODO(bogdando): implement podified FQDNs check for Nova endpoints
+- name: Check that Nova endpoints are defined and pointing to the podified FQDNs and that Nova API responds
+  ansible.builtin.shell: |
+    {{ shell_header }}
+    {{ oc_header }}
+    alias openstack="oc exec -t openstackclient -- openstack"
+
+    ${BASH_ALIASES[openstack]} endpoint list | grep nova
```
Comment (gibizer) [inline tests/roles/nova_adoption/tasks/main.yaml:83]: Does this check fails if nova is not in the list?

File: tests/roles/nova_adoption/tasks/main.yaml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    ${BASH_ALIASES[openstack]} server list
```
Comment (gibizer) [inline tests/roles/nova_adoption/tasks/main.yaml:84]: This does not assert that the test instance is in the returned list

File: tests/roles/nova_adoption/tasks/wait.yaml [hunk 8]
```diff
@@ -0,0 +1,35 @@
+# NOTE(bogdando): Status phase 'Running' doesn't necessarily mean it IS running in fact.
+# Instead, check for CRs status, then attempt exec'ing on the conductors pods to run live migrations,
+# with retries as guardrails for real running statuses of pods
+- name: wait for Nova control plane services' CRs to become ready
+  ansible.builtin.shell: |
+    {{ shell_header }}
+    {{ oc_header }}
+    oc get {{ service.cr }} --field-selector metadata.name={{ service.name }} -o jsonpath='{.items[0].status.conditions}' \
```
Comment (gibizer) [inline tests/roles/nova_adoption/tasks/wait.yaml:8]: oc wait --for condition=Ready --timeout=300s Nova/nova

File: tests/roles/nova_adoption/tasks/main.yaml (hunk 6 above)
Comment (bogdando) [reply tests/roles/nova_adoption/tasks/main.yaml:83]: yes, grep returns 1

File: tests/roles/nova_adoption/tasks/main.yaml (hunk 7 above)
Comment (bogdando) [reply tests/roles/nova_adoption/tasks/main.yaml:84]: the advanced checks can be found in another PR

File: tests/roles/nova_adoption/tasks/main.yaml (hunk 6 above)
Comment (gibizer) [reply tests/roles/nova_adoption/tasks/main.yaml:83]: Cool

File: tests/roles/nova_adoption/tasks/main.yaml (hunk 7 above)
Comment (gibizer) [reply tests/roles/nova_adoption/tasks/main.yaml:84]: OK

File: tests/roles/dataplane_adoption/tasks/main.yaml (hunk 5 above)
Comment (bogdando) [reply tests/roles/dataplane_adoption/tasks/main.yaml]: done

File: docs/openstack/edpm_adoption.md (hunk 4 above)
Comment (bogdando) [reply docs/openstack/edpm_adoption.md]: done

File: docs/openstack/nova_adoption.md (hunk 3 above)
Comment (bogdando) [reply docs/openstack/nova_adoption.md]: done

File: tests/roles/nova_adoption/tasks/wait.yaml [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# NOTE(bogdando): Status phase 'Running' doesn't necessarily mean it IS running in fact.
+# Instead, wait for CR Ready status, in short runs (https://github.com/kubernetes/kubectl/issues/1120)
+# of a total 10x6x5=300s waiting
```
Comment (gibizer) [inline tests/roles/nova_adoption/tasks/wait.yaml:3]: * We are waiting for Nova CR status not Pod status so I don't think this bug is relevant. Have you see this hanging?
* I think your math is wrong here. If I understand correctly ansible runs the shell command 6 times and waits 5 seconds between each run. As each oc run takes a maximum 10 seconds, the whole waiting time is (5 + 10) x 6 = 90 seconds not 300 seconds

Comment (bogdando) [reply tests/roles/nova_adoption/tasks/wait.yaml:3]: indeed, this issue probably have no sense for CRs

**Review by gibizer — APPROVED** [review]
My comments were fixed. If CI is green the I'm OK to land this.

**Comment by fao89** [comment]
```
TASK [osp.edpm.edpm_nova : Copy the migration ssh public key as authorized_keys to the nova user] ***
Wednesday 15 November 2023  14:48:56 +0000 (0:00:00.750)       0:00:24.214 **** 
[0;31mfatal: [standalone]: FAILED! => {"changed": false, "msg": "Source /var/lib/openstack/config/nova/ssh-publickey not found"}[0m
```
https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/f64c83e/controller/pod/nova.compute.extraconfig-openstack-8vq6r-logs.txt

**Comment by gibizer** [comment]
> ```
> TASK [osp.edpm.edpm_nova : Copy the migration ssh public key as authorized_keys to the nova user] ***
> Wednesday 15 November 2023  14:48:56 +0000 (0:00:00.750)       0:00:24.214 **** 
> �[0;31mfatal: [standalone]: FAILED! => {"changed": false, "msg": "Source /var/lib/openstack/config/nova/ssh-publickey not found"}�[0m
> ```
> 
> https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/f64c83e/controller/pod/nova.compute.extraconfig-openstack-8vq6r-logs.txt

Strange. We have the secret generated at https://github.com/openstack-k8s-operators/data-plane-adoption/pull/191/files#diff-0d721d5c31018c7d1dff0df6f5732b2d29a0cd1e814a5af5ce80d2d6c23c80d3R118 and that is included in the DataPlaneService at https://github.com/openstack-k8s-operators/data-plane-adoption/pull/191/files#diff-0d721d5c31018c7d1dff0df6f5732b2d29a0cd1e814a5af5ce80d2d6c23c80d3R153 and we using that server at https://github.com/openstack-k8s-operators/data-plane-adoption/pull/191/files#diff-0d721d5c31018c7d1dff0df6f5732b2d29a0cd1e814a5af5ce80d2d6c23c80d3R178

**Review by gibizer — COMMENTED** [review]
I see the reason of the CI failure now

**Comment by fao89** [comment]
> recheck

it will fail again, see @gibizer comments above

**Comment by bogdando** [comment]
recheck

**Comment by bogdando** [comment]
@jistr I couldn't get the CI failure cause https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/cc30154/controller/pod/dataplane-operator-controller-manager-569f9b45d-5qt4d-describe.txt

shall we just increase timeout? it takes more than 40m it seems?

File: tests/roles/dataplane_adoption/tasks/main.yaml [hunk 10]
```diff
@@ -108,60 +108,51 @@
...
+    apiVersion: dataplane.openstack.org/v1beta1
+    kind: OpenStackDataPlaneService
+    metadata:
+      name: nova-compute-extraconfig
+      namespace: openstack
+    spec:
+      label: nova.compute.extraconfig
+      configMaps:
+        - nova-compute-extraconfig
+      configSecrets:
```
[the remark this answers is not in the document]
Comment (fao89) [reply tests/roles/dataplane_adoption/tasks/main.yaml:151]: @bogdando ^

**Comment by fao89** [comment]
> @jistr I couldn't get the CI failure cause https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/cc30154/controller/pod/dataplane-operator-controller-manager-569f9b45d-5qt4d-describe.txt
> 
> shall we just increase timeout? it takes more than 40m it seems?

it is not a timeout issue,
it is a failure with the secret:
```
TASK [osp.edpm.edpm_nova : Copy the migration ssh public key as authorized_keys to the nova user] ***
Thursday 16 November 2023  18:56:20 +0000 (0:00:00.842)       0:00:25.837 ***** 
[0;31mfatal: [standalone]: FAILED! => {"changed": false, "msg": "Source /var/lib/openstack/config/nova/ssh-publickey not found"}[0m
```
https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/cc30154/controller/pod/nova.compute.extraconfig-openstack-tj68g-logs.txt

see: https://github.com/openstack-k8s-operators/data-plane-adoption/pull/191#discussion_r1394453454

File: tests/roles/dataplane_adoption/tasks/main.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (fao89) [reply tests/roles/dataplane_adoption/tasks/main.yaml:151]:
```suggestion
      secrets:
```

File: docs/openstack/edpm_adoption.md [hunk 11]
```diff
@@ -95,55 +95,57 @@ EOF
...
+    apiVersion: dataplane.openstack.org/v1beta1
+    kind: OpenStackDataPlaneService
+    metadata:
+      name: nova-compute-extraconfig
+      namespace: openstack
+    spec:
+      label: nova.compute.extraconfig
+      configMaps:
+        - nova-compute-extraconfig
+      configSecrets:
```
[the remark this answers is not in the document]
Comment (fao89) [reply docs/openstack/edpm_adoption.md:135]:
```suggestion
      secrets:
```

**Comment by fao89** [comment]
> @jistr I couldn't get the CI failure cause https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/cc30154/controller/pod/dataplane-operator-controller-manager-569f9b45d-5qt4d-describe.txt
> 
> shall we just increase timeout? it takes more than 40m it seems?

it is not a timeout issue,
it is a failure with the secret:
```
TASK [osp.edpm.edpm_nova : Copy the migration ssh public key as authorized_keys to the nova user] ***
Thursday 16 November 2023  18:56:20 +0000 (0:00:00.842)       0:00:25.837 ***** 
[0;31mfatal: [standalone]: FAILED! => {"changed": false, "msg": "Source /var/lib/openstack/config/nova/ssh-publickey not found"}[0m
```
https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/cc30154/controller/pod/nova.compute.extraconfig-openstack-tj68g-logs.txt

see: https://github.com/openstack-k8s-operators/data-plane-adoption/pull/191#discussion_r1394453454

**Comment by bogdando** [comment]
> > @jistr I couldn't get the CI failure cause https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/cc30154/controller/pod/dataplane-operator-controller-manager-569f9b45d-5qt4d-describe.txt
> > shall we just increase timeout? it takes more than 40m it seems?
> 
> it is not a timeout issue, it is a failure with the secret:
> 
> ```
> TASK [osp.edpm.edpm_nova : Copy the migration ssh public key as authorized_keys to the nova user] ***
> Thursday 16 November 2023  18:56:20 +0000 (0:00:00.842)       0:00:25.837 ***** 
> �[0;31mfatal: [standalone]: FAILED! => {"changed": false, "msg": "Source /var/lib/openstack/config/nova/ssh-publickey not found"}�[0m
> ```
> 
> https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/cc30154/controller/pod/nova.compute.extraconfig-openstack-tj68g-logs.txt
> 
> see: [#191 (comment)](https://github.com/openstack-k8s-operators/data-plane-adoption/pull/191#discussion_r1394453454)

thanks, I've read it back now :)
related question do you  think that deployment should fail, not timeout in such cases? looks like another related bug discovered (aside of my typo)

**Comment by fao89** [comment]
good point, we should make the deployment fail on these cases

**Comment by fao89** [comment]
I notice the ansible gets stuck on the task error:
https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/f64c83e/controller/pod/nova.compute.extraconfig-openstack-8vq6r-logs.txt
as the playbook doesn't finish to run, the AEE job keeps JobStatus as Running:
https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/f64c83e/controller/crs/openstackansibleees.yaml
and that's why the deployment doesn't get updated:
https://logserver.rdoproject.org/91/191/a5672bf71cfdb47b751a8752410fa0d0b5fe38b6/github-check/data-plane-adoption-github-rdo-centos-9-extracted-crc/f64c83e/controller/crs/openstackdataplanedeployments.yaml


So I'm trying: https://github.com/openstack-k8s-operators/data-plane-adoption/pull/201
to see if it unstucks ansible

**Pull request merged**